src/tools/rbd_wnbd/wnbd_wmi.cc

```diff
@@ -0,0 +1,219 @@
+  }
+
+  hres = CoInitializeSecurity(
+    NULL, -1, NULL, NULL,
+    RPC_C_AUTHN_LEVEL_DEFAULT,
+    RPC_C_IMP_LEVEL_IMPERSONATE,
+    NULL,
+    EOAC_NONE,
+    NULL);
+  if (FAILED(hres)) {
+    derr << "CoInitializeSecurity failed. HRESULT: " << hres << dendl;
+    CoUninitialize();
+    return hres;
+  }
+
+  return 0;
+}
+
+void WmiConnection::close()
+{
+  if (wbem_svc != NULL) {
+    wbem_svc->Release();
+    wbem_svc = NULL;
+  }
+  if (wbem_loc != NULL) {
+    wbem_loc->Release();
+    wbem_loc = NULL;
+  }
+}
+
+HRESULT WmiConnection::initialize()
+{
+  HRESULT hres = CoCreateInstance(
+    CLSID_WbemLocator, 0, CLSCTX_INPROC_SERVER,
+    IID_IWbemLocator, (LPVOID*)&wbem_loc);
+  if (FAILED(hres)) {
+    derr << "CoCreateInstance failed. HRESULT: " << hres << dendl;
+    goto Exit;
+  }
+
+  hres = wbem_loc->ConnectServer(
+    _bstr_t(ns.c_str()).GetBSTR(), NULL, NULL, NULL,
+    WBEM_FLAG_CONNECT_USE_MAX_WAIT, NULL, NULL,
+    &wbem_svc);
+  if (FAILED(hres)) {
+    derr << "Could not connect to WMI service. HRESULT: " << hres << dendl;
+    goto Exit;
+  }
+
+  hres = CoSetProxyBlanket(
+    wbem_svc, RPC_C_AUTHN_WINNT, RPC_C_AUTHZ_NONE, NULL,
+    RPC_C_AUTHN_LEVEL_CALL, RPC_C_IMP_LEVEL_IMPERSONATE, NULL, EOAC_NONE);
+  if (FAILED(hres)) {
+    derr << "CoSetProxyBlanket failed. HRESULT:" << hres << dendl;
+  }
+
+Exit:
+  if (FAILED(hres)) {
+    close();
+  }
+  return hres;
+}
+
+HRESULT get_property_str(
+  IWbemClassObject* cls_obj,
+  const std::wstring& property,
+  std::wstring& value)
+{
+  VARIANT vt_prop;
+  VariantInit(&vt_prop);
+  HRESULT hres = cls_obj->Get(property.c_str(), 0, &vt_prop, 0, 0);
+  if (!FAILED(hres)) {
+    VARIANT vt_bstr_prop;
+    VariantInit(&vt_bstr_prop);
+    hres = VariantChangeType(&vt_bstr_prop, &vt_prop, 0, VT_BSTR);
+    if (!FAILED(hres)) {
+      value = vt_bstr_prop.bstrVal;
+    }
+    VariantClear(&vt_bstr_prop);
+  }
+  VariantClear(&vt_prop);
+
+  if (FAILED(hres)) {
+    derr << "Could not get WMI property: " << to_string(property)
+         << ". HRESULT: " << hres << dendl;
+  }
+  return hres;
+}
+
+HRESULT get_property_int(
+  IWbemClassObject* cls_obj,
+  const std::wstring& property,
+  uint32_t& value)
+{
+  VARIANT vt_prop;
+  VariantInit(&vt_prop);
+  HRESULT hres = cls_obj->Get(property.c_str(), 0, &vt_prop, 0, 0);
+  if (!FAILED(hres)) {
+    VARIANT vt_bstr_prop;
+    VariantInit(&vt_bstr_prop);
+    hres = VariantChangeType(&vt_bstr_prop, &vt_prop, 0, VT_UINT);
+    if (!FAILED(hres)) {
+      value = vt_bstr_prop.intVal;
+    }
+    VariantClear(&vt_bstr_prop);
+  }
+  VariantClear(&vt_prop);
+
+  if (FAILED(hres)) {
+    derr << "Could not get WMI property: " << to_string(property)
+         << ". HRESULT: " << hres << dendl;
+  }
+  return hres;
+}
+
+HRESULT WmiSubscription::initialize()
+{
+  HRESULT hres = conn.initialize();
+  if (FAILED(hres)) {
+    derr << "Could not create WMI connection" << dendl;
+    return hres;
+  }
+
+  if (!conn.wbem_svc) {
```

Comment:
Is this possible?  `WmiConnection::initialize` passes `wbem_svc` pointer to `CoSetProxyBlanket`.  Can this API return success when passed a NULL `wbem_svc`?

Reply:
Good point, I think we can skip this check or move it above the CoSetProxyBlanket call.